File: tests/Tests.CSharp/ZetaId/BitLayoutAgreementTests.cs
using Xunit;
using Zeta.Core.CSharp.ZetaId;

namespace Zeta.Tests.CSharp.ZetaId;

/// <summary>
/// The V8-cycle cross-check, ported from Rust — because C# and F# did not have it.
///
/// <para><see cref="BitLayout.Create"/> has TWO independent construction paths,
/// <c>CreateTopDown</c> (the canonical authoring order, matching the human-readable layout spec) and
/// <c>CreateBottomUp</c>. They must produce identical field offsets; that redundancy IS the check, in
/// the same way the multi-oracle byte-lock is — two independent computations of one answer.</para>
///
/// <para>WHY THIS FILE EXISTS (2026-08-11). A mapping pass over the ZetaId bit layout found that only
/// RUST asserts the two paths agree (<c>src/Core.Rust.ZetaId/src/bit_layout.rs:200-203</c>). In C# and
/// F#, <c>BottomUp</c> appears in no test and no production call path — <c>Default</c> is
/// <c>TopDown</c> in both — so it is unreachable dead code. The consequence is a silent trap: a layout
/// edit applied to <c>CreateTopDown</c> and not to <c>CreateBottomUp</c> produces <b>no compile error
/// and no test failure</b>. The bottom-up path quietly retains the old field allocation and waits for
/// whoever next calls <c>Create(LayoutDirection.BottomUp)</c>.</para>
///
/// <para>TypeScript, Python and Go are not exposed — they build their layout once, directly from the
/// generated constants, so there is nothing to keep in sync.</para>
///
/// <para>This is a general safety net, not a one-off: it protects every future layout change, and it
/// was written <i>before</i> one — a bit-reclamation is under consideration — so the trap is closed
/// before the change that would spring it.</para>
/// </summary>
public class BitLayoutAgreementTests
{
    // NOTE — why there is no whole-object `Assert.Equal(td, bu)` here, unlike the F# sibling.
    //
    // `BitLayout` is a `sealed class` (BitLayout.cs:3) with no value-equality override, so
    // `Assert.Equal` on two instances compares REFERENCES. A first draft of this file asserted
    // whole-object equality and failed with two instances whose printed field values were
    // character-for-character identical — a false alarm that briefly looked like a real layout
    // divergence in production code.
    //
    // The F# sibling CAN compare whole records, because F# records get structural equality by
    // default and C# classes do not. Same intent, necessarily expressed differently per language —
    // which is itself the kind of cross-oracle asymmetry a byte-locked format has to stay careful
    // about. Field-by-field is the correct C# form, and it is complete: every field is compared.

    [Fact]
    public void TheTwoPathsAgreeFieldByField()
    {
        // A half-applied edit fails here rather than lying dormant until someone calls BottomUp,
        // and the failure names the field that drifted.
        var td = BitLayout.Create(LayoutDirection.TopDown);
        var bu = BitLayout.Create(LayoutDirection.BottomUp);

        Assert.Equal(td.Randomness, bu.Randomness);
        Assert.Equal(td.Location, bu.Location);
        Assert.Equal(td.Momentum, bu.Momentum);
        Assert.Equal(td.Persona, bu.Persona);
        Assert.Equal(td.Authority, bu.Authority);
        Assert.Equal(td.Category, bu.Category);
        Assert.Equal(td.Chromosome, bu.Chromosome);
        Assert.Equal(td.Timestamp, bu.Timestamp);
        Assert.Equal(td.Version, bu.Version);
    }
}

File: tests/Tests.FSharp/ZetaId/BitLayoutAgreement.Tests.fs
module Zeta.Tests.FSharp.ZetaId.BitLayoutAgreementTests

open global.Xunit
open Zeta.Core.FSharp.ZetaId

// ═══════════════════════════════════════════════════════════════════
// The V8-cycle cross-check, ported from Rust — because F# and C# did not have it.
//
// `BitLayout.create` has TWO independent construction paths, `createTopDown` (the canonical
// authoring order, matching the human-readable layout spec) and `createBottomUp`. They must produce
// identical field offsets; that redundancy IS the check, in exactly the way the four-oracle byte-lock
// is: two independent computations of the same answer.
//
// WHY THIS FILE EXISTS (2026-08-11). A mapping pass over the ZetaId bit layout found that only RUST
// asserts the two paths agree (`src/Core.Rust.ZetaId/src/bit_layout.rs:200-203`). In F# and C#,
// `BottomUp` appears in no test and no production call path — `Default` is `TopDown` in both — so it
// is unreachable dead code. The consequence is a silent trap: any layout edit applied to
// `createTopDown` and NOT to `createBottomUp` produces **no compile error and no test failure**. The
// bottom-up path simply retains the old field allocation and waits for whoever next calls
// `create BottomUp`.
//
// TypeScript, Python and Go are not exposed: they build their layout once, directly from the
// generated constants, so there is nothing to keep in sync.
//
// This is a general safety net rather than a one-off: it protects EVERY future layout change, and it
// was written before one — a bit-reclamation is under consideration — precisely so the trap is closed
// before the change that would spring it.
// ═══════════════════════════════════════════════════════════════════

[<Fact>]
let ``TopDown and BottomUp produce the identical layout — a half-applied edit is caught here`` () =
    // Structural equality on the record compares every field's offset and width at once, so a change
    // applied to one path and not the other fails immediately rather than lying dormant.
    Assert.Equal(BitLayout.create TopDown, BitLayout.create BottomUp)

[<Fact>]
let ``the default layout IS the TopDown one — the canonical path is the one production uses`` () =
    // Pins the claim in BitLayout.fs's own doc comment. If Default were ever switched to BottomUp,
    // the untested path would silently become the production path.
    Assert.Equal(BitLayout.Default, BitLayout.create TopDown)

[<Fact>]
let ``the two paths agree field-by-field — so a failure names WHICH field drifted`` () =
    // The structural comparison above fails as one opaque inequality. These name the culprit, which
    // is what you actually want at 2am during a wire-format change.
    let td = BitLayout.create TopDown
    let bu = BitLayout.create BottomUp

    let fields =
        [ "Randomness", td.Randomness, bu.Randomness
          "Location", td.Location, bu.Location
          "Momentum", td.Momentum, bu.Momentum
          "Persona", td.Persona, bu.Persona
          "Authority", td.Authority, bu.Authority
          "Firefly", td.Firefly, bu.Firefly
          "Category", td.Category, bu.Category
          "Chromosome", td.Chromosome, bu.Chromosome
          "Timestamp", td.Timestamp, bu.Timestamp
          "Version", td.Version, bu.Version ]

    for name, a, b in fields do
        // Parenthesised deliberately: in argument position F# parses a bare `a = b` as a NAMED
        // ARGUMENT rather than an equality test (error FS0691) — a quietly wrong-shaped assertion.
        Assert.True((a = b), sprintf "field %s disagrees between TopDown and BottomUp: %A vs %A" name a b)
